sway/lock.c

```diff
@@ -0,0 +1,179 @@
+	wl_list_remove(&lock_state->unlock.link);
+	wl_list_remove(&lock_state->abandon.link);
+
+	struct sway_seat *seat;
+	wl_list_for_each(seat, &server.input->seats, link) {
+		seat_set_exclusive_client(seat, NULL);
+		// copied from seat_set_focus_layer -- deduplicate?
+		struct sway_node *previous = seat_get_focus_inactive(seat, &root->node);
+		if (previous) {
+			// Hack to get seat to re-focus the return value of get_focus
+			seat_set_focus(seat, NULL);
+			seat_set_focus(seat, previous);
+		}
+	}
+
+
+	// redraw everything
+	for (int i = 0; i < root->outputs->length; ++i) {
+		struct sway_output *output = root->outputs->items[i];
+		output_damage_whole(output);
+	}
+}
+
+static void handle_abandon(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_manager *lock_state = server.session_lock->data;
+
+	sway_log(SWAY_INFO, "session lock abandoned");
+	lock_state->lock = NULL;
+
+	wl_list_remove(&lock_state->new_surface.link);
+	wl_list_remove(&lock_state->unlock.link);
+	wl_list_remove(&lock_state->abandon.link);
+
+	struct sway_seat *seat;
+	wl_list_for_each(seat, &server.input->seats, link) {
+		seat->exclusive_client = NULL;
+	}
+
+	// redraw everything
+	for (int i = 0; i < root->outputs->length; ++i) {
+		struct sway_output *output = root->outputs->items[i];
+		output_damage_whole(output);
+	}
+}
+
+static void handle_session_lock(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_manager *manager = wl_container_of(listener, manager, new_lock);
+	struct wlr_session_lock_v1 *lock = data;
+	struct wl_client *client = wl_resource_get_client(lock->resource);
+
+	if (manager->lock) {
+		wlr_session_lock_v1_destroy(lock);
+		return;
+	}
+
+	sway_log(SWAY_DEBUG, "session locked");
+	manager->locked = true;
+	manager->lock = lock;
+	lock->data = manager;
+
+	struct sway_seat *seat;
+	wl_list_for_each(seat, &server.input->seats, link) {
+		seat_set_exclusive_client(seat, client);
+	}
+
+	wl_signal_add(&lock->events.new_surface, &manager->new_surface);
+	wl_signal_add(&lock->events.unlock, &manager->unlock);
+	wl_signal_add(&lock->events.destroy, &manager->abandon);
+
+	wlr_session_lock_v1_send_locked(lock);
+
+	// redraw everything
+	for (int i = 0; i < root->outputs->length; ++i) {
+		struct sway_output *output = root->outputs->items[i];
+		output_damage_whole(output);
+	}
+}
+
+static void handle_session_lock_destroy(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_manager *manager = wl_container_of(listener, manager, manager_destroy);
+	wl_list_remove(&manager->new_lock.link);
+	wl_list_remove(&manager->manager_destroy.link);
+	free(manager);
+}
+
+struct wlr_session_lock_manager_v1 *sway_session_lock_manager_create(void) {
+	struct wlr_session_lock_manager_v1 *session_lock = wlr_session_lock_manager_v1_create(server.wl_display);
+	struct sway_session_lock_manager *lock_state = calloc(1, sizeof(*lock_state));
+	if (!lock_state) {
+		return NULL;
+	}
+
+	session_lock->data = lock_state;
```

Comment:
If the type of `server->session_lock` is changed as described above, we can drop this.